lib/semantics/check-omp-structure.cc

```diff
@@ -0,0 +1,268 @@
+// Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "check-omp-structure.h"
+#include "../parser/parse-tree.h"
+
+namespace Fortran::semantics {
+
+bool OmpStructureChecker::HasInvalidWorksharingNesting(
+    const parser::CharBlock &source, const OmpDirectiveSet &set) {
+  // set contains all the invalid closely nested directives
+  // for the given directive (`source` here)
+  if (CurrentDirectiveIsNested() && set.test(GetContext().currentDirectiveEnum)) {
```

Comment:
Can `CurrentDirectiveIsNested()` ever return false here?

Reply:
Yes, it will return false on the outmost directive.

Reply:
Could you add a test that would fail without this check?

Reply:
Yes, but right now the check for DO loop constructs hasn't been implemented, I can add a dummy check to invalidate nested `parallel` constructs:
```
void OmpStructureChecker::Enter(const parser::OpenMPBlockConstruct &x) {
   const auto &dir{std::get<parser::OmpBlockDirective>(x.t)};
+  if (std::holds_alternative<parser::OmpBlockDirective::Parallel>(dir.u)) {
+    HasInvalidWorksharingNesting(dir.source, {OmpDirective::PARALLEL});
+  }
   OmpContext ct{dir.source};
   ompContext_.push_back(ct);
 }
```

For the code below, I will get a segfault when compiling it without the `CurrentDirectiveIsNested()` because for the first `parallel` directive, when it calls the above `Enter` function, the stack is empty and the call to the second condition `set.test(GetContext().currentDirectiveEnum)` will introduce the segfault:
```
  !$omp parallel
  !$omp parallel
  do i=1, 100
     a = 2.0
  enddo
  !$omp end parallel
  !$omp end parallel
end program
```

Reply:
Addressed in [f10920a](f10920a83170b9350b7efd8a164f9b0f70ac46dc).